external-crates/move/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -140,84 +126,706 @@ struct PackageWithResolverTOML<'a>(&'a Package);
+        deps_digest: Option<String>,
     ) -> Result<DependencyGraph> {
-        let mut graph = DependencyGraph {
+        // collect sub-graphs for "regular" and "dev" dependencies
+        let mut dep_graphs = self.collect_graphs(
+            parent,
+            root_manifest.package.name,
+            root_path.clone(),
+            DependencyMode::Always,
+            &root_manifest.dependencies,
+        )?;
+        let mut dev_dep_graphs = self.collect_graphs(
+            parent,
+            root_manifest.package.name,
+            root_path.clone(),
+            DependencyMode::DevOnly,
+            &root_manifest.dev_dependencies,
+        )?;
+
+        let mut combined_graph = DependencyGraph {
             root_path: root_path.clone(),
-            root_package: root_package.package.name,
+            root_package: root_manifest.package.name,
             package_graph: DiGraphMap::new(),
             package_table: BTreeMap::new(),
             always_deps: BTreeSet::new(),
+            manifest_digest,
+            deps_digest,
         };
+        // ensure there's always a root node, even if it has no edges
+        combined_graph
+            .package_graph
+            .add_node(combined_graph.root_package);
 
-        // Ensure there's always a root node, even if it has no edges.
-        graph.package_graph.add_node(graph.root_package);
-        // Collect external resolution requests and process them later to check for "safe"
-        // overlapping packages existing in both externally and internally resolved graphs.
-        let mut external_requests = vec![];
-        graph
-            .extend_graph(
-                &PM::DependencyKind::default(),
-                root_package,
-                &root_path,
-                dependency_cache,
-                &mut external_requests,
-                &mut BTreeMap::new(),
-                progress_output,
-            )
-            .with_context(|| {
-                format!(
-                    "Failed to resolve dependencies for package '{}'",
-                    graph.root_package
-                )
-            })?;
+        // get overrides
+        let overrides = DependencyGraphBuilder::<Progress>::collect_overrides(
+            parent,
+            &root_manifest.dependencies,
+        )?;
+        let dev_overrides = DependencyGraphBuilder::<Progress>::collect_overrides(
+            parent,
+            &root_manifest.dev_dependencies,
+        )?;
 
-        for ExternalRequest {
-            mode,
-            from,
-            to,
-            resolver,
-            pkg_path,
-            overrides,
-        } in external_requests
-        {
-            graph
-                .resolve_externally(
-                    mode,
-                    from,
-                    to,
-                    resolver,
-                    &pkg_path,
+        for (_, (g, is_override)) in dep_graphs.iter_mut() {
+            if !*is_override {
```

Comment:
Good point. We have to do it carefully to avoid over-pruning by actually trying to prune the overridden package itself. Fixed and added a test for this.